kernel/src/hil/nfc.rs

```diff
@@ -0,0 +1,105 @@
+use crate::returncode::ReturnCode;
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+/// An enum to keep track of the NFC field status
+pub enum NfcFieldState {
+    /// Initial value that indicates no NFCT field events.
+    None,
+    /// The NFCT FIELDLOST event has been set.
+    On,
+    /// The NFCT FIELDDETECTED event has been set.
+    Off,
+    /// Both NFCT field events have been set - ambiguous state.
+    Unknown,
+}
+
+/// Controls an NFC tag main functionalities
+pub trait NfcTag<'a> {
+    /// Set the client instance that will handle callbacks
+    fn set_client(&self, client: &'a dyn Client<'a>);
+
+    /// Enable NFC sense field mode, and subscribe to the relevant interrupts.
+    /// Also set up the default configurations of how frame delay should be
+    /// dealt with (e.g. the maximum delay before a timeout).
+    /// This function should never fail.
+    fn enable(&self);
+
+    /// Enable tag emulation by triggering the necessary task.
+    /// This function should never fail.
+    fn activate(&self);
+    /// Notify the client by calling `field_lost()`. Then disable tag emulation
+    /// by triggering the necessary task. And go back to state of sense field.
+    /// This function should never fail.
+    fn deactivate(&self);
+
+    /// Pass the buffer to be transmitted and the amount of data and take
+    /// ownership of it. Subscribe to the relevant interrupt and trigger
+    /// the task for transmission.
+    ///     
+    /// On success returns the length of data to be sent.
+    /// On failure returns an error code and the buffer passed in.
+    fn transmit_buffer(
+        &self,
+        tx_buffer: &'static mut [u8],
+        tx_amount: usize,
+    ) -> Result<usize, (ReturnCode, &'static mut [u8])>;
+
+    /// Pass a buffer for receiving data and take ownership of it.
+    ///     
+    /// On success returns nothing.
+    /// On failure returns an error code and the buffer passed in.
+    fn receive_buffer(
+        &self,
+        rx_buffer: &'static mut [u8],
+    ) -> Result<(), (ReturnCode, &'static mut [u8])>;
+
+    /// Configuration of the Tag according to its Type.
+    ///
+    /// Return `SUCCESS` if the tag type is supported and
+    /// `ENOSUPPORT` otherwise.
+    fn configure(&self, tag_type: u8) -> ReturnCode;
+
+    /// Set the maximum frame delay in number of 13.56 MHz clocks.
+    fn set_framedelaymax(&self, max_delay: u32);
+
+    /// Enable the interrupt for event SELECTED
+    fn unmask_select(&self);
```

Comment:
I don't think I quite understand what this is, but it feels too chip-specific for a HIL